### PR TITLE
[core] fix installModules random crash

### PR DESCRIPTION
**Pull request by Kudo**
# Why

the `TypeError: Cannot read property 'NativeModule' of undefined` still happens occasionally. the problem is that installModules may happen right after the [block](https://github.com/facebook/react-native/blob/a5eeea814de57cd307a6625c6991f48b778d356b/packages/react-native/ReactAndroid/src/main/java/com/facebook/react/runtime/ReactHostImpl.java#L1008-L1055). the `JavaScriptContextHolder` may be ready after that on the [second block](https://github.com/facebook/react-native/blob/a5eeea814de57cd307a6625c6991f48b778d356b/packages/react-native/ReactAndroid/src/main/java/com/facebook/react/runtime/ReactHostImpl.java#L1060-L1109). we are not able to install expo global object when JavaScriptContextHolder is not ready 

# How

it's not clear to me whether it should be an issue from react native core. but to unblock our testing, let's add a busy waiting workaround first.

# Test Plan

```sh
$ git clone https://github.com/brentvatne/test-updates.git
$ cd test-updates
# remove expo-updates from package.json to make the repro minimal
$ yarn
$ npx expo run:android --variant release

# try the following command many times until the app stucking at white screen
$ adb shell am force-stop com.example.withnewarch && adb shell am start -n com.example.withnewarch/.MainActivity
```

# Checklist

- [x] Documentation is up to date to reflect these changes (eg: https://docs.expo.dev and README.md).
- [x] Conforms with the [Documentation Writing Style Guide](https://github.com/expo/expo/blob/main/guides/Expo%20Documentation%20Writing%20Style%20Guide.md)
- [x] This diff will work correctly for `npx expo prebuild` & EAS Build (eg: updated a module plugin).

### CODE REVIEW

File: packages/expo-modules-core/android/src/main/java/expo/modules/kotlin/ExpoBridgeModule.kt [hunk 1]
```diff
@@ -8,15 +9,33 @@ import java.lang.ref.WeakReference
...
+) :
+  ReactContextBaseJavaModule(reactContext) {
   override fun getName(): String = "ExpoModulesCore"
 
   @ReactMethod(isBlockingSynchronousMethod = true)
   fun installModules(): Boolean {
+    // Bridgeless ReactHostImpl may have BridgelessReactContext ready but not ReactInstance.
+    // Try to busy wait until ReactInstance is available so we could get the javaScriptContextHolder.
+    tryWaitSync(waitMs = 100, retries = 10) {
+      reactApplicationContext.hasActiveReactInstance()
```
Comment (lukmccall) [inline packages/expo-modules-core/android/src/main/java/expo/modules/kotlin/ExpoBridgeModule.kt:24]: Why can we run a js method when the application context isn't ready? It's looking like a bug in the React native core. Regardless, let's merge it for now, although I don't like solutions like this 😅

**Review by lukmccall — APPROVED** [review]

**Comment by lukmccall** [comment]
@Kudo, can you raise that problem in the workplace?

File: packages/expo-modules-core/android/src/main/java/expo/modules/kotlin/ExpoBridgeModule.kt (hunk 1 above)
Comment (Kudo) [reply packages/expo-modules-core/android/src/main/java/expo/modules/kotlin/ExpoBridgeModule.kt:24]: i think that's a bug inside ReactHostImpl where all [mReactInstanceTaskRef.get().getResult()](https://github.com/facebook/react-native/blob/99d0c540cac42efec2ab5943b8d0510bbf94e16b/packages/react-native/ReactAndroid/src/main/java/com/facebook/react/runtime/ReactHostImpl.java#L551-L573) return null. i would raise the bug to meta.
would merge this first and remove it once meta resolve the issue.

**Review by tsapeta — APPROVED** [review]

**Comment by sterlingwes** [comment]
I'm trying to trace down the status of this since we're encountering this issue even with this wait (RN74 and 75 on expo 51).

Are there any issues open on the RN side since you reached out to Meta?

**Comment by clmoreno** [comment]
I'm also encountering this issue frequently on release builds :(

**Comment by sterlingwes** [comment]
@clmoreno FWIW the fix for me was to backport all of the recent changes in `RNHeadlessAppLoader.kt` which didn't find their way into sdk 51 (we're running SDK 51 without new arch)

I didn't pinpoint the actual issue b/c it was so hard to reproduce, but I think it was the "[run on UI thread](https://github.com/expo/expo/blob/34a259f6dfea0b51acfb3deba4ebe4e481e31fd2/packages/expo-modules-core/android/src/main/java/expo/modules/adapters/react/apploader/RNHeadlessAppLoader.kt#L60)" enforcement

**Comment by clmoreno** [comment]
> @clmoreno FWIW the fix for me was to backport all of the recent changes in `RNHeadlessAppLoader.kt` which didn't find their way into sdk 51 (we're running SDK 51 without new arch)
> 
> I didn't pinpoint the actual issue b/c it was so hard to reproduce, but I think it was the "[run on UI thread](https://github.com/expo/expo/blob/34a259f6dfea0b51acfb3deba4ebe4e481e31fd2/packages/expo-modules-core/android/src/main/java/expo/modules/adapters/react/apploader/RNHeadlessAppLoader.kt#L60)" enforcement

I've been testing this with some beta users, it looks like is working!!


Here is the patch file i created if someone needs it.
File name: `expo-modules-core+1.12.26.patch`
https://www.npmjs.com/package/patch-package

```git
diff --git a/node_modules/expo-modules-core/android/src/main/java/expo/modules/adapters/react/apploader/RNHeadlessAppLoader.kt b/node_modules/expo-modules-core/android/src/main/java/expo/modules/adapters/react/apploader/RNHeadlessAppLoader.kt
index 2719497..fe213f1 100644
--- a/node_modules/expo-modules-core/android/src/main/java/expo/modules/adapters/react/apploader/RNHeadlessAppLoader.kt
+++ b/node_modules/expo-modules-core/android/src/main/java/expo/modules/adapters/react/apploader/RNHeadlessAppLoader.kt
@@ -10,7 +10,7 @@ import expo.modules.apploader.HeadlessAppLoader
 import expo.modules.core.interfaces.Consumer
 import expo.modules.core.interfaces.DoNotStrip
 
-private val appRecords: MutableMap<String, ReactInstanceManager> = mutableMapOf()
+private val appRecords: MutableMap<String, ReactContext> = mutableMapOf()
 
 class RNHeadlessAppLoader @DoNotStrip constructor(private val context: Context) : HeadlessAppLoader {
 
@@ -27,13 +27,13 @@ class RNHeadlessAppLoader @DoNotStrip constructor(private val context: Context)
         reactInstanceManager.addReactInstanceEventListener(object : ReactInstanceEventListener {
           override fun onReactContextInitialized(context: ReactContext) {
             HeadlessAppLoaderNotifier.notifyAppLoaded(params.appScopeKey)
+            reactInstanceManager.removeReactInstanceEventListener(this)
+            appRecords[params.appScopeKey] = context
             callback?.apply(true)
           }
         })
-        appRecords[params.appScopeKey] = reactInstanceManager
-        if (reactInstanceManager.hasStartedCreatingInitialContext()) {
-          reactInstanceManager.recreateReactContextInBackground()
-        } else {
+        // Ensure that we're starting the react host on the main thread
+        android.os.Handler(context.mainLooper).post {
           reactInstanceManager.createReactContextInBackground()
         }
       } else {
@@ -46,16 +46,20 @@ class RNHeadlessAppLoader @DoNotStrip constructor(private val context: Context)
 
   override fun invalidateApp(appScopeKey: String?): Boolean {
     return if (appRecords.containsKey(appScopeKey) && appRecords[appScopeKey] != null) {
-      val appRecord: ReactInstanceManager = appRecords[appScopeKey]!!
-      android.os.Handler(context.mainLooper).post {
-        // Only destroy the `ReactInstanceManager` if it does not bind with an Activity.
-        // And The Activity would take over the ownership of `ReactInstanceManager`.
-        // This case happens when a user clicks a background task triggered notification immediately.
-        if (appRecord.lifecycleState == LifecycleState.BEFORE_CREATE) {
-          appRecord.destroy()
+      val reactContext = appRecords[appScopeKey] ?: return false
+      val reactNativeHost = (reactContext.applicationContext as ReactApplication).reactNativeHost
+      if (reactNativeHost.hasInstance()) {
+        val reactInstanceManager: ReactInstanceManager = reactNativeHost.reactInstanceManager
+        android.os.Handler(reactContext.mainLooper).post {
+          // Only destroy the `ReactInstanceManager` if it does not bind with an Activity.
+          // And The Activity would take over the ownership of `ReactInstanceManager`.
+          // This case happens when a user clicks a background task triggered notification immediately.
+          if (reactInstanceManager.lifecycleState == LifecycleState.BEFORE_CREATE) {
+            reactInstanceManager.destroy()
+          }
+          HeadlessAppLoaderNotifier.notifyAppDestroyed(appScopeKey)
+          appRecords.remove(appScopeKey)
         }
-        HeadlessAppLoaderNotifier.notifyAppDestroyed(appScopeKey)
-        appRecords.remove(appScopeKey)
       }
       true
     } else {
@@ -63,8 +67,11 @@ class RNHeadlessAppLoader @DoNotStrip constructor(private val context: Context)
     }
   }
 
-  override fun isRunning(appScopeKey: String?): Boolean =
-    appRecords.contains(appScopeKey) && appRecords[appScopeKey]!!.hasStartedCreatingInitialContext()
+  override fun isRunning(appScopeKey: String?): Boolean  {
+    val reactContext = appRecords[appScopeKey] ?: return false
+    val reactNativeHost = (reactContext.applicationContext as ReactApplication).reactNativeHost
+    return reactNativeHost.reactInstanceManager.hasStartedCreatingInitialContext()
+  }
 
   //endregion HeadlessAppLoader
 }

```

**Pull request merged**